### ISSUE TITLE
An evaluation of (cdr '(10 (20 30) 40)) should output ((20 30) 40)

### ISSUE THREAD

**Issue by kariya-mitsuru** [mattn/cisp#19]
In the current implementation, an evaluation of `(cdr '(10 (20 30) 40))` outputs `(20 30)`.

Sorry, I cannot fix this problem because the current implementation is too complex to fix for me.

**Comment by mattn** [comment]
could you please explain how this behave?

**Comment by kariya-mitsuru** [comment]
The samples below show common implementation's structure.
Their `car` and `cdr` only get the `car` ptr and `cdr` ptr.
Their `cons` only makes a cons cell with its first argument as `car` ptr and its second argument as `cdr` ptr.

```
 (1 . 2)    |           (1 2)                  |           (1 2 . 3)
            |                                  |
+---+---+   |   +---+---+    +---+---+         |   +---+---+    +---+---+
|car|cdr|   |   |car|cdr| -> |car|cdr|         |   |car|cdr| -> |car|cdr|
+---+---+   |   +---+---+    +---+---+         |   +---+---+    +---+---+
  |   |     |     |            |   |           |     |            |   |
  v   v     |     v            v   v           |     v            v   v
 +-+ +-+    |    +-+          +-+ nil(NULL)    |    +-+          +-+ +-+
 |1| |2|    |    |1|          |2|              |    |1|          |2| |3|
 +-+ +-+    |    +-+          +-+              |    +-+          +-+ +-+


 ((10 . 20) . (30 . 40))    |           (1 2 3)
                            |
    +---+---+               |   +---+---+    +---+---+    +---+---+
    |car|cdr|               |   |car|cdr| -> |car|cdr| -> |car|cdr|
    +---+---+               |   +---+---+    +---+---+    +---+---+
      |   |                 |     |            |            |   |
  +---+   +----+            |     v            v            v   v
  |            |            |    +-+          +-+          +-+ nil(NULL)
  V            v            |    |1|          |2|          |3|
+---+---+    +---+---+      |    +-+          +-+          +-+
|car|cdr|    |car|cdr|      |
+---+---+    +---+---+      |
  |   |        |   |        |
  v   v        v   v        |
+--+ +--+    +--+ +--+      |
|10| |20|    |30| |40|      |
+--+ +--+    +--+ +--+      |


(10 (20 30) 40))

+---+---+    +---+---+    +---+---+
|car|cdr| -> |car|cdr| -> |car|cdr|
+---+---+    +---+---+    +---+---+
  |            |            |   |
  v            v            v   v
+--+         +---+---+    +--+ nil(NULL)
|10|         |car|cdr|    |40|
+--+         +---+---+    +--+
               |   |
               v   v
             +--+ +---+---+
             |20| |car|cdr|
             +--+ +---+---+
                    |   |
                    v   v
                  +--+ nil(NULL)
                  |30|
                  +--+
```

**Comment by kariya-mitsuru** [comment]
It's also worth noting that `'a` is abbreviation for `(quote a)`, so `(car ''a)` should be evaluated to `quote`.

**Comment by mattn** [comment]
Great! Thanks.

Current implementation is similar to the same you mentioned.

![](http://go-gyazo.appspot.com/2a9bc2221a6addba.png)

This is hacky to easy to handle pointer to move car to cdr by holding values on car/cdr like below.

```
(10 (20 30) 40))

+-------+---+    +---+---+    +-------+--------+
|car(10)|cdr| -> |car|cdr| -> |car(40)|cdr(nil)|
+-------+---+    +---+---+    +-------+--------+
                   |
                   v
                 +-------+---+
                 |car(20)|cdr|
                 +-------+---+
                       |
                       v
                      +-------+--------+
                      |car(30)|cdr(nil)|
                      +-------+--------+
```

**Comment by mattn** [comment]
So I'm thinking the fixing will be easy. However, this may be ugly.

**Comment by kariya-mitsuru** [comment]
TL;DR

I may understand current structure.
I think that `(10 (20 30) 40)` needs one more cell.

```
+---+--------+
|car|cdr(nil)|
+---+--------+
  |
  v
+-------+---+    +---+---+    +-------+--------+
|car(10)|cdr| -> |car|cdr| -> |car(40)|cdr(nil)|
+-------+---+    +---+---+    +-------+--------+
                   |
                   v
                 +-------+---+    +-------+--------+
                 |car(20)|cdr| -> |car(30)|cdr(nil)|
                 +-------+---+    +-------+--------+
```

And `((10 . 20) . (30 . 40))` has a structure like below.

```
+---+--------+
|car|cdr(nil)|
+---+--------+
  |
  v
+---+---+    +---+--------+
|car|cdr| -> |car|cdr(nil)|
+---+---+    +---+--------+
  |            |
  |            v
  |          +-------+---+         +---+--------+
  |          |car(30)|cdr| ------> |car|cdr(nil)|
  |          +-------+---+         +---+--------+
  |                                  |
  v                                  v
+-------+---+    +---+--------+    +-------+--------+
|car(10)|cdr| -> |car|cdr(nil)|    |car(40)|cdr(nil)|
+-------+---+    +---+--------+    +-------+--------+
                   |
                   v
                 +-------+--------+
                 |car(20)|cdr(nil)|
                 +-------+--------+
```

So, `(10 (20))` has a structure like below.

```
+---+--------+
|car|cdr(nil)|
+---+--------+
  |
  v
+-------+---+    +---+--------+
|car(10)|cdr| -> |car|cdr(nil)|
+-------+---+    +---+--------+
                   |
                   v
                 +-------+--------+
                 |car(20)|cdr(nil)|
                 +-------+--------+
```

But it is identical to `(10 . 20)`.

**Comment by mattn** [comment]
Yes, perfect. And I forgot to put the first leading cell in my explain. There is already exists in the top of tree.

**Comment by kariya-mitsuru** [comment]
Do you plan to distinguish `(10 (20))` from `(10 . 20)` ?

**Comment by mattn** [comment]
> Do you plan to distinguish (10 (20)) from (10 . 20) ?

Sorry, I'm not master of lisp. Could you please point a URL or something of the spec about this?
- Why it's same
- How the implementation should do about this

**Comment by kariya-mitsuru** [comment]
I'm sorry I didn't make it clear enough.

In the current implementation, a representation of `(10 (20))` is the structure that I explained above, but a representation of `(10 . 20)` is also same structure, unfortunately.
In fact, if I entered the `'(10 (20))` into the current implementation, it outputs `(10 . 20)`.

`(10 (20))` is completely different from `(10 . 20)` for LISP, so I think they should have different representation each other.

cf. general representation of `(10 . 20)`, `(10 (20))` and `(10 20)`.

```
(10 . 20)   |   (10 (20))                     |   (10 20)
            |                                 |
+---+---+   |   +---+---+    +---+--------+   |   +---+---+    +---+--------+
|car|cdr|   |   |car|cdr| -> |car|cdr(nil)|   |   |car|cdr| -> |car|cdr(nil)|
+---+---+   |   +---+---+    +---+--------+   |   +---+---+    +---+--------+
  |   |     |     |            |              |     |            |
  v   v     |     v            v              |     v            v
+--+ +--+   |   +--+         +---+--------+   |   +--+         +--+
|10| |20|   |   |10|         |car|cdr(nil)|   |   |10|         |20|
+--+ +--+   |   +--+         +---+--------+   |   +--+         +--+
                               |
                               v
                             +--+
                             |20|
                             +--+
```

**Comment by mattn** [comment]
last change should handle this. https://github.com/mattn/cisp/commit/c3f39d489a08cb7d21d9712829dd9a77a375addb

![](http://go-gyazo.appspot.com/eba2b327ea5982e9.png)

**Comment by mattn** [comment]
But test seems to be fail.

**Comment by mattn** [comment]
Hmmm, seems broken. So reverted last commit. I must look into it again.

**Comment by mattn** [comment]
However, I can fix this because I got https://github.com/mattn/cisp/issues/19#issuecomment-232381864

Maybe, parsing should start at whether the root object or start of the list of operations like progn.

**Comment by mattn** [comment]
Now pushed changes of data structure for the `(10 (20))` into master branch. but work in progress.

```
01-basic.lisp: OK
02-defun.lisp: OK
03-dotimes.lisp: OK
04-comment.lisp: OK
05-car.lisp: OK
06-cdr.lisp: OK
07-mod.lisp: OK
08-setq.lisp: OK
09-load.lisp: OK
10-length.lisp: OK
11-apply.lisp: OK
12-eq.lisp: OK
13-quote.lisp: OK
14-type-of.lisp: OK
15-cons.lisp: OK
16-rest.lisp: OK
17-lambda.lisp: SKIP
18-funcall.lisp: SKIP
19-concatenate.lisp: OK
20-make-string.lisp: OK
21-let.lisp: SKIP
22-rplaca.lisp: OK
23-rplacd.lisp: OK
24-evenp.lisp: OK
25-oddp.lisp: OK
```

**Comment by kariya-mitsuru** [comment]
The new structure(964d652) resolves the problem mentioned above but I think that it brings another problem.

If my understanding is correct, `(10 . (20))` has the structure below in the new implementation.

```
+----+--+---+--------+
|CELL|  |car|cdr(nil)|
+----+--+---+--------+
          |
          v
+----+--+--------+---+    +----+--+---+--------+
|INT |10|car(nil)|cdr| -> |CELL|  |car|cdr(nil)|
+----+--+--------+---+    +----+--+---+--------+
                                    |
                                    v
                          +----+--+---+--------+
                          |CELL|  |car|cdr(nil)|
                          +----+--+---+--------+
                                    |
                                    v
                          +----+--+--------+--------+
                          |INT |20|car(nil)|cdr(nil)|
                          +----+--+--------+--------+
```

But I think that `(10 (20))` also has the same structure undesirably.

Note that `(10 . (20))` should have the same structure as `(10 20)`, rather than `(10 (20))`.

**Comment by mattn** [comment]
Hmm, looks like your said.
 ![](http://go-gyazo.appspot.com/49a053df8fd4cfbc.png)

**Comment by mattn** [comment]
I'm still not understand well.

`(A . B)` should be handled as `(A . B)` but
`(A . (B))` should be handled as `(A B)` and
`(A . ((B)))` should be handled as `(A (B))`. Right?

**Comment by kariya-mitsuru** [comment]
Yes, right.

`(A . B)` represents that it is a cons cell, with that its car points `A` and its cdr points `B`.
On the other hand, `(B)` is an abbreviation of `(B . nil)`, so it represents that it is a cons cell, with that its car points `B` and its cdr points `nil` (i.e. points none).

```
 (A . B)    |   (B)  =  (B . nil)
            |
+---+---+   |   +---+---+
|car|cdr|   |   |car|cdr|
+---+---+   |   +---+---+
  |   |     |     |   |
  v   v     |     v   v
 +-+ +-+    |    +-+ nil(NULL)
 |A| |B|    |    |B|
 +-+ +-+    |    +-+
```

Additionally, `(A B . C)` is an abbreviation of `(A . (B . C))`, and `(A B)` is an abbreviation of `(A . (B . nil))`.

```
(A B . C)  =  (A . (B . C))   |   (A B)  =  (A . (B . nil))  =  (A . (B))
                              |
+---+---+                     |   +---+---+ <--------- When this cons cell is focused,
|car|cdr|                     |   |car|cdr|            its car points A and
+---+---+                     |   +---+---+            its cdr points (B . nil) = (B),
  |   |                       |     |   |              so it can be represented as (A . (B)).
  v   v                       |     v   v
 +-+ +---+---+                |    +-+ +---+---+ <---- When this cons cell is focused,
 |A| |car|cdr|                |    |A| |car|cdr|       it seems (B . nil) = (B).
 +-+ +---+---+                |    +-+ +---+---+
       |   |                  |          |   |
       v   v                  |          v   v
      +-+ +-+                 |         +-+ nil(NULL)
      |B| |C|                 |         |B|
      +-+ +-+                 |         +-+
```

Similarly,  `(A B C D . E)` is an abbreviation of `(A . (B . (C . (D . E))))`, and `(A B C D)` is an abbreviation of `(A . (B . (C . (D . nil))))`.

```
(A B C D . E)  =  (A . (B . (C . (D . E))))   |   (A B C D)  =  (A . (B . (C . (D . nil))))
                                              |
+---+---+                                     |   +---+---+
|car|cdr|                                     |   |car|cdr|
+---+---+                                     |   +---+---+
  |   |                                       |     |   |
  v   v                                       |     v   v
 +-+ +---+---+                                |    +-+ +---+---+
 |A| |car|cdr|                                |    |A| |car|cdr|
 +-+ +---+---+                                |    +-+ +---+---+
       |   |                                  |          |   |
       v   v                                  |          v   v
      +-+ +---+---+                           |         +-+ +---+---+
      |B| |car|cdr|                           |         |B| |car|cdr|
      +-+ +---+---+                           |         +-+ +---+---+
            |   |                             |               |   |
            v   v                             |               v   v
           +-+ +---+---+                      |              +-+ +---+---+
           |C| |car|cdr|                      |              |C| |car|cdr|
           +-+ +---+---+                      |              +-+ +---+---+
                 |   |                        |                    |   |
                 v   v                        |                    v   v
                +-+ +-+                       |                   +-+ nil(NULL)
                |D| |E|                       |                   |D|
                +-+ +-+                       |                   +-+
```

That's all.

So, `(A (B))` turns into `(A . ((B)))` as follows.

```
(A (B))  =  (A . ((B) . nil))  =  (A . ((B . nil) . nil))

+---+---+ <--------- When this cons cell is focused,
|car|cdr|            its car points A and
+---+---+            its cdr points ((B)),
  |   |              so it can be represented as (A . ((B))).
  v   v
 +-+ +---+---+ <---- When this cons cell is focused,
 |A| |car|cdr|       its car points (B . nil) = (B) and
 +-+ +---+---+       its cdr points nil (NULL),
       |   |         so it seems ((B . nil) . nil) = ((B) . nil) = ((B)).
       |   v
       |  nil(NULL)
       v
     +---+---+ <---- When this cons cell is focused,
     |car|cdr|       it seems (B . nil) = (B).
     +---+---+
       |   |
       v   v
      +-+ nil(NULL)
      |B|
      +-+
```

**Comment by mattn** [comment]
I have a question. What is a condition whether cisp should display something as `(A . B)` or `(A B)` ?

**Comment by kariya-mitsuru** [comment]
In general, LISP prefers list notation over dotted pair notation whenever possible.

So, `(A)` is preferred over `(A . nil)`, `(A B)` is preferred over `(A . (B . nil))` or `(A . (B))`.

It is achieved by printing a cons-cell like below.
1. Print '`(`'(left-parenthesis).
2. Print the car-part of the cons-cell.
   (Of course, if the car-part is another cons-cell, print recursively.)
3. If the cdr-part of the cons-cell points to another cons-cell,
   then print '` `'(space), set the cons-cell that the cdr-part points to as a current cons-cell and go to step 2.
4. If the cdr-part of the cons-cell is not `nil` (i.e. not NULL),
   then print '`.`'(space-dot-space), and print the cdr-part.
5. Print '`)`'(right-parenthesis).

**Comment by mattn** [comment]
Thanks for your kindness!

**Comment by mattn** [comment]
誤解が無いように日本語で。
えーっと

```
(10 . 20)   |   (10 (20))                     |   (10 20)
            |                                 |
+---+---+   |   +---+---+    +---+--------+   |   +---+---+    +---+--------+
|car|cdr|   |   |car|cdr| -> |car|cdr(nil)|   |   |car|cdr| -> |car|cdr(nil)|
+---+---+   |   +---+---+    +---+--------+   |   +---+---+    +---+--------+
  |   |     |     |            |              |     |            |
  v   v     |     v            v              |     v            v
+--+ +--+   |   +--+         +---+--------+   |   +--+         +--+
|10| |20|   |   |10|         |car|cdr(nil)|   |   |10|         |20|
+--+ +--+   |   +--+         +---+--------+   |   +--+         +--+
                               |
                               v
                             +--+
                             |20|
                             +--+
```

これを見る限り、`.` を出すかどうかは今見てるセルの cdr が別の car に繋がっていない場合という条件だとまずいでしょうか？

**Comment by kariya-mitsuru** [comment]
OK です。
上の処理フローで言うと、3. の条件の部分に当たります。
すみません、処理フローの 3. 分かりにくいですね、ちょっと修正しました。

**Comment by mattn** [comment]
ようやくわかりました。今の構造の良くないところは `(1 2)` を cdr で繋げてしまっている点ですね。

https://github.com/mattn/cisp/issues/19#issuecomment-233839510

これの左と右が逆になっている。結構大がかりそう。。。

**Comment by mattn** [comment]
テストを足したいのですが以下で良いでしょうか？足らない物とかあるでしょうか？

t/26-conscell.lisp

```
(print (cdr '(10 (20 30) 40)))
(print '((20 30) 40))
(print '(1 . 2))
(print '(1 2))
(print '(1 2 . 3))
(print '((10 . 20) . (30 . 40)))
(print '(1 2 3))
(print '(10 . 20))
(print '(10 (20)))
(print '(A . B))
(print '(B))
(print '(B . nil))
(print '(A B))
(print '(A . (B . nil)))
(print '(A . (B)))
(print '(A (B)))
(print '(A . ((B) . nil)))
(print '(A . ((B . nil) . nil)))
```

t/26-conscell.out

```
((20 30) 40)
((20 30) 40)
(1 . 2)
(1 2)
(1 2 . 3)
((10 . 20) 30 . 40)
(1 2 3)
(10 . 20)
(10 (20))
(A . B)
(B)
(B)
(A B)
(A B)
(A B)
(A (B))
(A (B))
(A (B))
```

**Comment by kariya-mitsuru** [comment]
他の関数等と組み合わせないのであれば、このあたりはいかがでしょうか。
入力

```
(print '(10 . (20 . 30)))
(print '(10 (20 . 30)))
(print '())
(print '(10 . ()))
(print '(() . ()))
(print '(() () ()))
(print '(nil . nil))
(print '(nil nil nil))
(print '(nil . 10))
(print '(nil 10))
```

出力

```
(10 20 . 30)
(10 (20 . 30))
nil
(10)
(nil)
(nil nil nil)
(nil)
(nil nil nil)
(nil . 10)
(nil 10)
```

**Comment by mattn** [comment]
ありがとうございます。
fix-structure ブランチで構造をちゃんと直す修正をしていてほぼデータ構造は↑の通りに入ったのですが表示がよく分からなくて

``` lisp
(print '((10 . 20) . (30 . 40)))
```

これが

``` lisp
((10 . 20) 30 . 40)
```

こうなるって部分が良く分かってないです。現状の表示部分はこんな感じです。

``` c
static void
print_cell(size_t nbuf, char *buf, NODE *node, int mode) {
  NODE *c;
  strncat(buf, "(", nbuf);
  c = node->car;

  while (c) {
    if (c != node->car) strncat(buf, " ", nbuf);
    if (c->car) print_node(nbuf, buf, c->car, mode);
    if (c->cdr && c->cdr->t != NODE_CELL) {
      strncat(buf, " . ", nbuf);
      print_node(nbuf, buf, c->cdr, mode);
      break;
    }
    c = c->cdr;
  }

  strncat(buf, ")", nbuf);
}
```

**Comment by mattn** [comment]
あーわかった。こうですかね。

``` c
static void
print_cell(size_t nbuf, char *buf, NODE *node, int mode) {
  NODE *c;
  strncat(buf, "(", nbuf);
  c = node->car;

  while (c) {
    if (c != node->car) strncat(buf, " ", nbuf);
    if (c->car) print_node(nbuf, buf, c->car, mode);
    if (c->cdr) {
      if (c->cdr->t != NODE_CELL) {
        strncat(buf, " . ", nbuf);
        print_node(nbuf, buf, c->cdr, mode);
        break;
      }
      c = c->cdr->car; // ココがミソ
      continue;
    }
    c = c->cdr;
  }

  strncat(buf, ")", nbuf);
}
```

**Comment by mattn** [comment]
ちがうー。

**Comment by kariya-mitsuru** [comment]
今の構造だと、子要素が cons-cell の場合に、普通の LISP よりも階層が一つ深くなってしまってるんですよね…

追記
例えば、`'(10 . 20)` の場合、最上位の cons-cell の car が更に cons-cell を指していて、その cons-cell が `10` と `20` を指す構造になってますよね。

**Comment by kariya-mitsuru** [comment]
`parse_paren` の

```
    } else {
      x = new_node();
      x->t = NODE_CELL;
      x->car = child;
      if (!head->car) {
        head->car = x;
      } else {
        node->cdr = x;
      }
      node = x;
    }
```

の部分を、

```
    } else {
      if (head->car) {
        x = new_node();
        x->t = NODE_CELL;
        node->cdr = x;
        node = x;
      }
      node->car = child;
    }
```

のようにすると、普通っぽい構造になりますね。
ただし、リストの最初の要素か否かを `head->car` で判断しているところは、`nil` を本物の `NULL` に変更した場合に問題になってしまいますが…

**Comment by mattn** [comment]
えとコード群

```
(xxx)
(yyy)
```

を束ねる list (progn)が一番上にいてそれの事かと思います。

**Comment by mattn** [comment]
確かに cons cell のネストが深いすね。

**Comment by kariya-mitsuru** [comment]
cons cell のネストが普通の場合にはこんな感じでしょうか…

```
static void
print_cell(size_t nbuf, char *buf, NODE *node, int mode) {
  strncat(buf, "(", nbuf);

  while (node) {
    if (node->car)
      print_node(nbuf, buf, node->car, mode);
    else
      strncat(buf, "nil", nbuf);
    if (!node->cdr || node->cdr->t == NODE_NIL)
      break;
    if (node->cdr->t != NODE_CELL) {
      strncat(buf, " . ", nbuf);
      print_node(nbuf, buf, node->cdr, mode);
      break;
    }
    strncat(buf, " ", nbuf);
    node = node->cdr;
  }

  strncat(buf, ")", nbuf);
}
```

**Comment by mattn** [comment]
あざます！ :bow:

**Comment by kariya-mitsuru** [comment]
蛇足ですが、`parse_any` や `parse_paren` 等の戻り値を `NODE *` 型にして、オブジェクトの生成を関数内にしたほうが、`nil` や `()` をパースした際に `NULL` を返却できるので `nil` の扱いがラクになる気がします…（`nil == NULL` と実装する）
今返却している `char *` は `SCANNER` を実装するとエラー状態の返却も含めて引数だけで事足りるような気がするので…（`SCANNER` は目的が違うかな？）

**Comment by mattn** [comment]
> 蛇足ですが、parse_any や parse_paren 等の戻り値を NODE \* 型にして、オブジェクトの生成を関数内にしたほうが、nil や () をパースした際に NULL を返却できるので nil の扱いがラクになる気がします…（nil == NULL と実装する）

はい。ひとまず基本の動きが出来たらやろうかと思います。

> 返却している char \* は SCANNER を実装するとエラー状態の返却も含めて引数だけで事足りるような気がするので…（SCANNER は目的が違うかな？）

そこ悩んでる所です。SCANNER に読み取りエラーとか行番号とかその辺を持たせて返す感じにしようかと検討中です。

```
01-basic.lisp: OK
02-defun.lisp: OK
03-dotimes.lisp: OK
04-comment.lisp: OK
05-car.lisp: OK
06-cdr.lisp: OK
07-mod.lisp: OK
08-setq.lisp: OK
09-load.lisp: OK
10-length.lisp: OK
11-apply.lisp: SKIP
12-eq.lisp: OK
13-quote.lisp: OK
14-type-of.lisp: OK
15-cons.lisp: OK
16-rest.lisp: SKIP
17-lambda.lisp: SKIP
18-funcall.lisp: OK
19-concatenate.lisp: OK
20-make-string.lisp: OK
21-let.lisp: OK
22-rplaca.lisp: OK
23-rplacd.lisp: OK
24-evenp.lisp: OK
25-oddp.lisp: OK
26-conscell.lisp: SKIP
```

あとちょいという感じです。

**Comment by mattn** [comment]
新構造でテストがパスしたのでこちら閉じます。長々とかつ親切にありがとうございました。
